File: android/src/main/kotlin/io/agora/agora_rtc_engine/AgoraRtcChannelPlugin.kt
package io.agora.agora_rtc_engine

import android.os.Handler
import android.os.Looper
import android.util.Log
import androidx.annotation.NonNull
import io.agora.iris.rtc.IrisRtcEngine
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

private class ApiTypeChannelCallApiMethodCallHandler(
  irisRtcEngine: IrisRtcEngine
) : CallApiMethodCallHandler(irisRtcEngine) {
  override fun callApi(apiType: Int, params: String?, sb: StringBuffer): Int {
    return irisRtcEngine.channel.callApi(apiType, params, sb)
  }

  override fun callApiWithBuffer(apiType: Int, params: String?, buffer: ByteArray?, sb: StringBuffer): Int {
    return irisRtcEngine.channel.callApi(apiType, params, buffer, sb)
  }
}

/** AgoraRtcChannelPlugin */
class AgoraRtcChannelPlugin(
//  private val rtcEnginePlugin: AgoraRtcEnginePlugin
  private val irisRtcEngine: IrisRtcEngine
) : FlutterPlugin, MethodCallHandler, EventChannel.StreamHandler {
  /// The MethodChannel that will the communication between Flutter and native Android
  ///
  /// This local reference serves to register the plugin with the Flutter Engine and unregister it
  /// when the Flutter Engine is detached from the Activity
  private lateinit var methodChannel: MethodChannel
  private lateinit var eventChannel: EventChannel
  private var eventSink: EventChannel.EventSink? = null
//  private val manager = RtcChannelManager { methodName, data -> emit(methodName, data) }
  private val handler = Handler(Looper.getMainLooper())

  private lateinit var callApiMethodCallHandler: CallApiMethodCallHandler

  fun initPlugin(binaryMessenger: BinaryMessenger) {
    methodChannel = MethodChannel(binaryMessenger, "agora_rtc_channel")
    methodChannel.setMethodCallHandler(this)
    eventChannel = EventChannel(binaryMessenger, "agora_rtc_channel/events")
    eventChannel.setStreamHandler(this)

    callApiMethodCallHandler = ApiTypeChannelCallApiMethodCallHandler(irisRtcEngine)


  }

  override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    initPlugin(binding.binaryMessenger)
  }

  override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    methodChannel.setMethodCallHandler(null)
    eventChannel.setStreamHandler(null)
  }

  override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {

    eventSink = events

    irisRtcEngine.channel.setEventHandler(EventHandler(eventSink))
    Log.e("MainActivity", "channel onListen: $eventSink")
  }

  override fun onCancel(arguments: Any?) {
    irisRtcEngine.channel.setEventHandler(null)
    eventSink = null
  }

  override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
    // Iris supported
    callApiMethodCallHandler.onMethodCall(call, result)
  }
}

File: macos/Classes/AgoraRtcChannelPlugin.swift
//
//  AgoraRtcChannelPlugin.swift
//  agora_rtc_engine
//
//  Created by LXH on 2020/6/29.
//

import Foundation

public class AgoraRtcChannelPlugin: NSObject, FlutterPlugin, FlutterStreamHandler {
    private final weak var rtcEnginePlugin: SwiftAgoraRtcEnginePlugin?
    private var methodChannel: FlutterMethodChannel?
    private var eventChannel: FlutterEventChannel?
    private var eventSink: FlutterEventSink?
    private lazy var manager: RtcChannelManager = {
        RtcChannelManager { [weak self] methodName, data in
            self?.emit(methodName, data)
        }
    }()

    init(_ rtcEnginePlugin: SwiftAgoraRtcEnginePlugin) {
        self.rtcEnginePlugin = rtcEnginePlugin
    }

    public static func register(with _: FlutterPluginRegistrar) {}

    public func initPlugin(_ registrar: FlutterPluginRegistrar) {
        methodChannel = FlutterMethodChannel(name: "agora_rtc_channel", binaryMessenger: registrar.messenger)
        eventChannel = FlutterEventChannel(name: "agora_rtc_channel/events", binaryMessenger: registrar.messenger)
        registrar.addMethodCallDelegate(self, channel: methodChannel!)
        eventChannel?.setStreamHandler(self)
    }

    public func detachFromEngine(for _: FlutterPluginRegistrar) {
        methodChannel?.setMethodCallHandler(nil)
        eventChannel?.setStreamHandler(nil)
        manager.Release()
    }

    public func onListen(withArguments _: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        eventSink = events
        return nil
    }

    public func onCancel(withArguments _: Any?) -> FlutterError? {
        eventSink = nil
        return nil
    }

    private func emit(_ methodName: String, _ data: [String: Any?]?) {
        var event: [String: Any?] = ["methodName": methodName]
        if let data = data {
            event.merge(data) { current, _ in
                current
            }
        }
        eventSink?(event)
    }

    private weak var engine: AgoraRtcEngineKit? {
        return rtcEnginePlugin?.engine
    }

    func channel(_ channelId: String) -> AgoraRtcChannel? {
        return manager[channelId]
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        if let params = call.arguments as? NSDictionary {
            let selector = NSSelectorFromString(call.method + "::")
            if manager.responds(to: selector) {
                if call.method == "create" {
                    params.setValue(engine, forKey: "engine")
                }
                manager.perform(selector, with: params, with: ResultCallback(result))
                return
            }
        } else {
            let selector = NSSelectorFromString(call.method + ":")
            if manager.responds(to: selector) {
                manager.perform(selector, with: ResultCallback(result))
                return
            }
        }
        result(FlutterMethodNotImplemented)
    }
}
